### ISSUE TITLE
retrieve a tuple directly in a for comprehension with ZIO?

### ISSUE THREAD

**Issue by alvNa** [zio/zio#2761]
Hi ZIO team,

I'm using ZIO library in several projects and I've checked that when you use ZIO in a for comprehension is not possible to retrieve a tuple directly by defining all it's fields, or at least I don't know how to do it. 

Example:

Without ZIO:
for {
      (a,b) <- Some(1,2)
} yield (a,b)

With ZIO:
    for {
      (a,b) <- ZIO.fromOption(Some(1,2))
    } yield (a,b)

If I do this the compiler complains. 

Obviously I can retrieve the tuple using a single parameter and access to the elements with p._1 p._2 etc.
    for {
      p <- ZIO.fromOption(Some(1,2))
    } yield p

Is there a way to do this ?

**Comment by ajaychandran** [comment]
You can with [better-monadic-for](https://github.com/oleg-py/better-monadic-for).

**Comment by alvNa** [comment]
Thank you for the information @ajaychandran.
But my intention is to achieve this with ZIO library without using Cats.

**Comment by adamgfraser** [comment]
Unfortunately this doesn't work today. To do a destructuring assignment in a for comprehension like this we need to implement `withFilter`:

```scala
trait ZIO[-R, +E, +A] {
  def withFilter(f: A => Boolean): ZIO[R, E, A]
}
```

The challenge is how to do this in a polymorphic way. We can `flatMap` to look at the `A` value and if the predicate is true we can just succeed with the `A` value. But what do we do if the predicate isn't true? We don't have an `E` value and `E` is completely polymorphic so we can't just make up an `E` value to fail with.

The current solution ZIO uses is defining `withFilter` like this:

```scala
def withFilter(predicate: A => Boolean)(implicit ev: NoSuchElementException <:< E): ZIO[R, E, A] =
  self.flatMap { a =>
    if (predicate(a)) ZIO.succeedNow(a)
    else ZIO.failNow(new NoSuchElementException("The value doesn't satisfy the predicate"))
  }
```

Basically we require `E` to be a super type of `NoSuchElementException` like `Throwable`, so you can do what you want with something like `Task`. But in your case `fromOption` returns `Unit` which is not related to `Throwable`.

So the options as I see it are:
1. Keep the current behavior.
2. Don't require `E` to be a super type of `NoSuchElementException`. This would make examples like this compile but the error type could be widened to `Any`, which is less than ideal.
3. Change `withFilter` to require the `E` type to be `Unit` and fail with `()`. This would in a way be more idiomatic since there is only one failure case here (the predicate not being satisfied) and it would make this example compile but it would make other examples with `Throwable` no longer compile. Maybe we could have overloaded variants but that doesn't seem very scalable and appears to create issues with duplicate method signatures.
4. Die instead of fail when the predicate is not satisfied. This would be completely polymorphic but failing with an unchecked versus a checked exception here seems highly problematic.
5. Require implicit evidence of some kind of "default" failure for a type `E` and then use that for the filter failure case, but that would require a type class which we have tried to avoid and doesn't really seem worth it for this one case.

This is all a long way of saying I very much agree that it would be nice to support doing this but I'm not sure how we actually can with a polymorphic error type. I even wonder to a certain extent whether we should support the existing version given it only works for a particular error type.

**Comment by ajaychandran** [comment]
@alvNa [better-monadic-for](https://github.com/oleg-py/better-monadic-for) is a compiler plugin that desugars `for` comprehensions for data-types like `ZIO` that do not define `withFilter`.
All you need to do is update your build file,
```sbt
addCompilerPlugin("com.olegpy" %% "better-monadic-for" % "0.3.1")
```
and your snippet will compile just fine.

**Comment by ajaychandran** [comment]
@adamgfraser 

When using the plugin, type inference works as expected.

```scala
  def task1: ZIO[Any, RuntimeException, Any] = ZIO.fail(new RuntimeException)
  def task2: ZIO[Any, Exception, Any] = ZIO.fail(new Exception)
  def task3: ZIO[Any, Unit, Any] = ZIO.fail(())

  def task12: ZIO[Any, Exception, Int] =
    for {
      _ <- task1
      _ <- task2
      (l, r) <- zio.ZIO.effectTotal(1 -> 1)
    } yield l + r

  def task13: ZIO[Any, Any, Unit] =
    for {
      _ <- task12
      _ <- task3
    } yield ()

```

**Comment by adamgfraser** [comment]
@ajaychandran Yes understood. The plugin is one solution for limitations in the Scala compiler's treatment of irrefutable destructurings. I think it is definitely good for people to know about the plugin and what it can do if they want to use it but I think there is still an issue of what the best treatment of this issue is in ZIO since ideally we would not like users to have to use plugins for desired behavior.

**Comment by ajaychandran** [comment]
@adamgfraser :+1:

@alvNa  For reference, the compiler issue is being tracked [here](https://github.com/lampepfl/dotty/issues/2573).

**Comment by alvNa** [comment]
I appreciate your explanation @adamgfraser. 
As you mentioned it would be an nice feature, but I understand it's not an easy task currently.
Thank you very much.

**Comment by adamgfraser** [comment]
Maybe we remove the existing `withFilter` and move it to ZIO Syntax where we can use some minimal `HasDefault[E]` implicit evidence to do this in a more polymorphic way.

@jdegoes What do you think?

**Comment by carlosedp** [comment]
Hi all reviving an old thread, this will be "fixed" in Scala 3.4 where tuples can be directly unpacked in the for comprehension.

The PR https://github.com/lampepfl/dotty/pull/18842 have been merged last week.

I did a sample at https://gist.github.com/carlosedp/c65b80ac1c0ddecd266f49248916a3df
